### PR TITLE
Add ZipBuilder support for tvOS and macOS

### PR DESCRIPTION
**Pull request by paulb777**
- Refactor to support regenerating pods for each platform
- Distinguish Platform (ios, tvos, macos) from TargetPlatform(iosDevice, tvosSimulator, etc)
- Include minimum supported version as part of Platform
- Add flags to support different platforms for different pods in Firebase distro (not supported for non-Firebase distros)
- Eliminate unused caching in `buildFramework`

Open question - Should all platforms be the default or only iOS?

TODO in subsequent PRs on the `zip-v3` branch:
- Improve code organization
- Add a `--skip-catalyst` option
- Reenable cron only zip tests

#no-changelog

### CODE REVIEW

File: ReleaseTooling/Sources/ZipBuilder/CocoaPodUtils.swift [hunk 1]
```diff
@@ -36,13 +36,40 @@ enum CocoaPodUtils {
...
+
+    init(name: String,
+         version: String?,
+         platforms: Set<String> = ["ios", "macos", "tvos"]) {
+      self.name = name
+      self.version = version
+      self.platforms = platforms
+    }
+
+    init(from decoder: Decoder) throws {
```
Comment (ryanwilson) [inline ReleaseTooling/Sources/ZipBuilder/CocoaPodUtils.swift:58]: Can we not get this for free from the compiler?

File: ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift [hunk 2]
```diff
@@ -702,19 +655,16 @@ struct FrameworkBuilder {
     return true
   }
 
-  /// Packages an XCFramework based on an almost complete framework folder (missing the binary but
-  /// includes everything else needed) and thin archives for each architecture slice.
+  /// Groups slices for each platform into a minimal set of frameworks
```
Comment (ryanwilson) [inline ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift:658]: Nit: add period at the end.

File: ReleaseTooling/Sources/ZipBuilder/ZipBuilder.swift [hunk 3]
```diff
@@ -666,78 +715,43 @@ struct ZipBuilder {
...
-    } else {
-      return (toInstall, nil)
+      frameworks.append(zipLocation)
+
+      CarthageUtils.generatePlistContents(
+        forName: framework.lastPathComponent.components(separatedBy: ".").first!,
+        withVersion: podInfo.version,
+        to: carthageLocation
+      )
+      carthageFrameworks.append(carthageLocation)
```
Comment (ryanwilson) [inline ReleaseTooling/Sources/ZipBuilder/ZipBuilder.swift:753]: Does this fail if carthage build is disabled, or does it return the plist contents?

File: ReleaseTooling/Sources/ZipBuilder/main.swift [hunk 4]
```diff
@@ -105,14 +104,23 @@ struct ZipBuilderTool: ParsableCommand {
           help: ArgumentHelp("The minimum supported iOS version."))
   var minimumIOSVersion: String
 
-  /// The list of architectures to build for.
+  /// The minimum macOS Version to build for.
+  @Option(default: "10.12",
```
Comment (ryanwilson) [inline ReleaseTooling/Sources/ZipBuilder/main.swift:108]: Nit: This and the one below may be able to fit on one line

File: ReleaseTooling/Sources/ZipBuilder/main.swift [hunk 5]
```diff
@@ -105,14 +104,23 @@ struct ZipBuilderTool: ParsableCommand {
...
+          help: ArgumentHelp("The minimum supported macOS version."))
+  var minimumMACOSVersion: String
```
Comment (ryanwilson) [inline ReleaseTooling/Sources/ZipBuilder/main.swift:110]: After looking at it further, I think this and the others should be renamed to `minimumMacOSVersion`

File: ReleaseTooling/Sources/ZipBuilder/main.swift [hunk 6]
```diff
@@ -228,19 +241,11 @@ struct ZipBuilderTool: ParsableCommand {
...
                              dynamicFrameworks: dynamic,
                              customSpecRepos: customSpecRepos)
-    let projectDir = FileManager.default.temporaryDirectory(withName: "project")
-
-    // If it exists, remove it before we re-create it. This is simpler than removing all objects.
-    if FileManager.default.directoryExists(at: projectDir) {
-      try FileManager.default.removeItem(at: projectDir)
-    }
-
-    CocoaPodUtils.podInstallPrepare(inProjectDir: projectDir, paths: paths)
```
Comment (ryanwilson) [inline ReleaseTooling/Sources/ZipBuilder/main.swift:243]: What's the reason for removing this?

File: ReleaseTooling/Sources/ZipBuilder/main.swift [hunk 7]
```diff
@@ -296,13 +299,13 @@ struct ZipBuilderTool: ParsableCommand {
...
 
-      FirebaseBuilder(zipBuilder: builder).build(in: projectDir,
-                                                 minimumIOSVersion: minimumIOSVersion,
+      FirebaseBuilder(zipBuilder: builder).build(templateDir: paths.templateDir,
                                                  carthageBuildOptions: carthageOptions)
     }
 
     if !keepBuildArtifacts {
-      FileManager.default.removeIfExists(at: projectDir.deletingLastPathComponent())
+      let tempDir = FileManager.default.temporaryDirectory(withName: "random")
```
Comment (ryanwilson) [inline ReleaseTooling/Sources/ZipBuilder/main.swift:307]: Should this be "random" or something else?

File: ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift (hunk 2 above)
Comment (paulb777) [reply ReleaseTooling/Sources/ZipBuilder/FrameworkBuilder.swift:658]: Done

File: ReleaseTooling/Sources/ZipBuilder/main.swift (hunk 4 above)
Comment (paulb777) [reply ReleaseTooling/Sources/ZipBuilder/main.swift:108]: Done.

File: ReleaseTooling/Sources/ZipBuilder/main.swift (hunk 5 above)
Comment (paulb777) [reply ReleaseTooling/Sources/ZipBuilder/main.swift:110]: Done.

File: ReleaseTooling/Sources/ZipBuilder/main.swift (hunk 6 above)
Comment (paulb777) [reply ReleaseTooling/Sources/ZipBuilder/main.swift:243]: There is now a separate projectDir for each platform.

File: ReleaseTooling/Sources/ZipBuilder/main.swift (hunk 7 above)
Comment (paulb777) [reply ReleaseTooling/Sources/ZipBuilder/main.swift:307]: Renamed to "placeholder". It seems like there should b a better way to delete the tempDir without generating the name, but I couldn't find it.

File: ReleaseTooling/Sources/ZipBuilder/ZipBuilder.swift (hunk 3 above)
Comment (paulb777) [reply ReleaseTooling/Sources/ZipBuilder/ZipBuilder.swift:753]: This is the binary only path and the return value gets ignored by the caller if Carthage is disabled.

**Pull request merged**